pkg/server/api/bundle/v1/service.go

```diff
@@ -161,17 +155,85 @@ func (s *Service) GetFederatedBundle(ctx context.Context, req *bundle.GetFederat
 		return nil, status.Errorf(codes.NotFound, "bundle for %q not found", req.TrustDomain)
 	}
 
-	b, err := applyMask(dsResp.Bundle, req.OutputMask)
+	b, err := api.BundleToProto(dsResp.Bundle)
 	if err != nil {
-		log.Errorf("Failed to apply mask: %v", err)
-		return nil, status.Errorf(codes.Internal, "failed to apply mask: %v", err)
+		log.WithError(err).Error("Failed to convert bundle")
+		return nil, status.Errorf(codes.Internal, "failed to convert bundle: %v", err)
 	}
 
+	applyBundleMask(b, req.OutputMask)
+
 	return b, nil
 }
 
 func (s *Service) BatchCreateFederatedBundle(ctx context.Context, req *bundle.BatchCreateFederatedBundleRequest) (*bundle.BatchCreateFederatedBundleResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method BatchCreateFederatedBundle not implemented")
+	log := rpccontext.Logger(ctx)
+
+	results := []*bundle.BatchCreateFederatedBundleResponse_Result{}
+	for _, b := range req.Bundle {
+		td, err := spiffeid.TrustDomainFromString(b.TrustDomain)
+		if err != nil {
+			log.Errorf("Trust domain argument is not a valid SPIFFE ID: %q", b.TrustDomain)
+			results = append(results, &bundle.BatchCreateFederatedBundleResponse_Result{
+				Status: CreateStatus(codes.InvalidArgument, "trust domain argument is not a valid SPIFFE ID: %q", b.TrustDomain),
+			})
+			continue
+		}
+
+		if s.td.Compare(td) == 0 {
+			log.Errorf("%q is this server own trust domain", td.String())
+			results = append(results, &bundle.BatchCreateFederatedBundleResponse_Result{
+				Status: CreateStatus(codes.InvalidArgument, "%q is this server own trust domain", td.String()),
+			})
+			continue
+		}
+
+		dsBundle, err := api.ProtoToBundle(b)
+		if err != nil {
+			log.WithError(err).Error("Unable to convert bundle")
+			results = append(results, &bundle.BatchCreateFederatedBundleResponse_Result{
+				Status: CreateStatus(codes.Internal, "unable to convert bundle: %v", err),
+			})
+			continue
+		}
+		resp, err := s.ds.CreateBundle(ctx, &datastore.CreateBundleRequest{
```

Comment:
I think before we call create bundle we must validate that rootca and public keys are valid